source/dol/bases/d_a_iceball.cpp

```diff
@@ -0,0 +1,506 @@
+        deleteRequest();
+    }
+}
+
+void daIceBall_c::initializeState_Move() {
+    static const float cs_speed_x[] = {3.0f, -3.0f};
+    static const float cs_max_speed_x[] = {1.5f, -1.5f};
+
+    float playerXSpeed = daPyMng_c::getPlayer(mPlayerNo)->mSpeed.x;
+    float moveBoostThreshold = sGlobalData_c<daIceBall_c>::mData.moveBoostThreshold;
+    float moveBoostScale = sGlobalData_c<daIceBall_c>::mData.mMoveBoostScale;
+
+    float speedBoost;
+    if (playerXSpeed > moveBoostThreshold) {
+        float baseBoost = moveBoostScale * moveBoostThreshold;
+        speedBoost = baseBoost + sGlobalData_c<daIceBall_c>::mData.calcExtraBoost(playerXSpeed - moveBoostThreshold);
+    } else if (playerXSpeed < -moveBoostThreshold) {
+        float baseBoost = -moveBoostScale * moveBoostThreshold;
+        speedBoost = baseBoost + sGlobalData_c<daIceBall_c>::mData.calcExtraBoost(playerXSpeed + moveBoostThreshold);
+    } else {
+        speedBoost = moveBoostScale * playerXSpeed;
+    }
+
+    mSpeed.set(speedBoost + cs_speed_x[mDirection], 3.7f, 0.0f);
+    mAccelY = -0.15f;
+    mSpeedMax.x = cs_max_speed_x[mDirection];
+    mMaxFallSpeed = -3.0f;
+    mAccelF = 0.04f;
+}
+
+void daIceBall_c::finalizeState_Move() {}
+
+void daIceBall_c::executeState_Move() {
+    EffectManager_c::SetIceBallEffect(&mPos);
+
+    if (mAliveTimer != 0 && --mAliveTimer == 0 && sm_IceBallAliveCount[mPlayerNo] > 0) {
+        sm_IceBallAliveCount[mPlayerNo]--;
+    }
+
+    calcSpeedX();
+    calcFallSpeed();
+    posMove();
+    chgZpos();
+
+    if (waterlineCheck()) {
+        setDeleteEffect();
+        deleteRequest();
+        return;
+    }
+
+    if (bgCheck()) {
+        dAudio::SoundEffectID_t(SE_OBJ_PNGN_ICEBALL_DISAPP).playMapSound(mPos, 0);
+        setDeleteEffect();
+        deleteRequest();
+        return;
+    }
+
+    if (mRc.getRide()) {
+        dAudio::SoundEffectID_t(SE_OBJ_PNGN_ICEBALL_DISAPP).playMapSound(mPos, 0);
+        setDeleteEffect();
+        deleteRequest();
+        return;
+    }
+
+    if (cullCheck()) {
+        deleteRequest();
+    }
+}
+
+void daIceBall_c::initializeState_EatIn() {}
+
+void daIceBall_c::finalizeState_EatIn() {}
+
+void daIceBall_c::executeState_EatIn() {
+    EffectManager_c::SetIceBallEffect(&mPos);
+    if (mEatState == EAT_STATE_EATEN) {
+        mStateMgr.changeState(StateID_EatNow);
+    }
+}
+
+void daIceBall_c::initializeState_EatNow() {}
+
+void daIceBall_c::finalizeState_EatNow() {}
+
+void daIceBall_c::executeState_EatNow() {}
+
+void daIceBall_c::initializeState_Kill() {
+    removeCc();
+    deleteRequest();
+}
+
+void daIceBall_c::finalizeState_Kill() {}
+
+void daIceBall_c::executeState_Kill() {}
+
+bool daIceBall_c::CheckIceballLimit(int playerId, int limitMode) {
+    if (daIceBall_c::sm_IceBallCount[playerId] < 6) {
+        if (limitMode == 1) {
+            return true;
+        }
+
+        if (daIceBall_c::sm_IceBallAliveCount[playerId] < 2) {
+            return true;
+        }
+    }
```

Comment:
Possible to have these limits as game constants?

Reply:
I'll make them static const class members, I think it makes more sense to do it like that for something that's only used in one class.